### ISSUE TITLE
Possible Performance Problems

### ISSUE THREAD

**Issue by UndeadZeratul** [stanhebben/MineTweaker3#132]
Recently our private server modpack has been hit with quite a bit of a performance issue.  It seems something in our scripts has caused us to have very slow load times when starting up a new SSP world or even connecting to our server.  [Here's](https://github.com/Nincraft/NincraftElectricBoogaloo2RevengeoftheFallen/issues/149) the issue on our side, it mentions which scripts are taking the longest from looking at the logs and seeing which ones it gets stuck on the longest.  [Here's](https://github.com/Nincraft/NincraftElectricBoogaloo2RevengeoftheFallen/tree/develop/scripts) the most recent version of our scripts, and as of writing this, is the version which has the issue.

Is there any method call that you know of that might have less-than-optimal performance?  Are we doing something dumb in our scripts at all?  The only thing that's common in all but 1 of the files that take a while to load, is when I try to unify a few ore dictionary entries, by doing an addAll() and a mirror() call.  Should I not be doing this? is there a better option, or should I just not try unifying ore dictionary entries (yet, if 3.0.10 or 3.1.0 improves these)?

Any other tips you can provide would be helpful.  I notice a lot of other scripts don't declare variables and instead just reference things directly in the recipe methods.  Is that a more optimal way of doing things, or does that not matter?  I'd like to include at least this many scripts in our pack, and in fact we're not done with the tweaking we'd like to implement, but if it's going to cause inherent performance problems, we might have to re-approach this some other way.

**Comment by stanhebben** [comment]
Having a quick look at the scripts, I can't see anything that looks specifically problematic. Recipes.remove is always the slowest, though, as it has to run through every recipe in the game in order to find the ones it has to remove - in a game with many mods (and thus, recipes) it could slow down loading. I wouldn't expect it to take minutes, but I haven't done any performance tests.

Unifying ore dictionaries is the most "hacky" part of MineTweaker, but they themselves are fast operations. (If there is an issue, it would most likely be a machine not picking up the unification)

I don't think there's anything specific to be done from your part. There is a lot of room for optimization in MineTweaker and I'm afraid it will be up to me to find and fix the slow parts.

**Comment by UndeadZeratul** [comment]
Ah, well that's somewhat good to hear, I thought I might have been doing something dumb that would be causing the slowness.

Is there something I can do with the recipe removals that can help?  Considering we do have GregTech installed, it's not surprising that if it loops through every recipe to find one that it can remove, that it'd take a while to do.

I am looking forward to 3.0.10 and 3.1.0 to see what kind of performance improvements we'll get.

**Comment by stanhebben** [comment]
Is there an easy way to install your modpack? It would be a nice test case.

**Comment by UndeadZeratul** [comment]
Oh, sure!  It's hosted privately on the Technic Launcher, but [here's](http://www.technicpack.net/modpack/details/nincraft-electric-boogaloo-17) the modpack webpage URL.  Bear in mind that because it's a large pack, the startup times will take a while (roughly 5 minutes or so on our computers) thanks to GregTech, and starting up a single player world, or even connecting to our server will end up taking about 2 minutes.

**Comment by UndeadZeratul** [comment]
So after doing some testing (via disabling a lot of GregTech - and I mean over 50,000 - recipes to shorten the list that MineTweaker would have to sift through) there wasn't any speed improvement.  The pack still takes ~5 minutes to start up, and ~2 minutes to log into a world.

However, I reverted all of my changes in the GT config file as well as uncommented my recipes.remove() methods in my scripts so that I was back to square one, and instead commented out every OreDict.addAll() and OreDict.mirror() call, and, well, see for yourself:

    [19:21:59] [Server thread/INFO] [STDOUT]: [minetweaker.runtime.MTTweaker:load:106]: Loading scripts
    [19:21:59] [Server thread/INFO] [STDOUT]: [minetweaker.runtime.MTTweaker:load:152]: MineTweaker: Loading NEB2 Big Reactors Tweaks.zs
    [19:21:59] [Server thread/INFO] [STDOUT]: [minetweaker.runtime.MTTweaker:load:152]: MineTweaker: Loading NEB2 Blood Magic Tweaks.zs
    [19:21:59] [Server thread/INFO] [STDOUT]: [minetweaker.runtime.MTTweaker:load:152]: MineTweaker: Loading NEB2 Botania Tweaks.zs
    [19:21:59] [Server thread/INFO] [STDOUT]: [minetweaker.runtime.MTTweaker:load:152]: MineTweaker: Loading NEB2 Carpenters Blocks Tweaks.zs
    [19:21:59] [Server thread/INFO] [STDOUT]: [minetweaker.runtime.MTTweaker:load:152]: MineTweaker: Loading NEB2 EnderIO Tweaks.zs
    [19:21:59] [Server thread/INFO] [STDOUT]: [minetweaker.runtime.MTTweaker:load:152]: MineTweaker: Loading NEB2 EnderStorage Tweaks.zs
    [19:22:00] [Server thread/INFO] [STDOUT]: [minetweaker.runtime.MTTweaker:load:152]: MineTweaker: Loading NEB2 ExU Tweaks.zs
    [19:22:00] [Server thread/INFO] [STDOUT]: [minetweaker.runtime.MTTweaker:load:152]: MineTweaker: Loading NEB2 GC3 Tweaks.zs
    [19:22:00] [Server thread/INFO] [STDOUT]: [minetweaker.runtime.MTTweaker:load:152]: MineTweaker: Loading NEB2 GregTech Tweaks.zs
    [19:22:00] [Server thread/INFO] [STDOUT]: [minetweaker.runtime.MTTweaker:load:152]: MineTweaker: Loading NEB2 Harvestcraft Tweaks.zs
    [19:22:00] [Server thread/INFO] [STDOUT]: [minetweaker.runtime.MTTweaker:load:152]: MineTweaker: Loading NEB2 Magical Crops Tweaks.zs
    [19:22:00] [Server thread/INFO] [STDOUT]: [minetweaker.runtime.MTTweaker:load:152]: MineTweaker: Loading NEB2 Mekanism Tweaks.zs
    [19:22:00] [Server thread/INFO] [STDOUT]: [minetweaker.runtime.MTTweaker:load:152]: MineTweaker: Loading NEB2 MFR Tweaks.zs
    [19:22:00] [Server thread/INFO] [STDOUT]: [minetweaker.runtime.MTTweaker:load:152]: MineTweaker: Loading NEB2 Nincrafty Things Tweaks.zs
    [19:22:00] [Server thread/INFO] [STDOUT]: [minetweaker.runtime.MTTweaker:load:152]: MineTweaker: Loading NEB2 PneumaticCraft Tweaks.zs
    [19:22:00] [Server thread/INFO] [STDOUT]: [minetweaker.runtime.MTTweaker:load:152]: MineTweaker: Loading NEB2 TiC Tweaks.zs
    [19:22:00] [Server thread/INFO] [STDOUT]: [minetweaker.runtime.MTTweaker:load:152]: MineTweaker: Loading NEB2 Torcherino Tweaks.zs
    [19:22:00] [Server thread/INFO] [STDOUT]: [minetweaker.runtime.MTTweaker:load:152]: MineTweaker: Loading NEB2 Vanilla Tweaks.zs
    [19:22:00] [Server thread/INFO] [STDOUT]: [minetweaker.runtime.MTTweaker:load:152]: MineTweaker: Loading PhoenixConfig AE2 Tweaks.zs
    [19:22:00] [Server thread/INFO] [STDOUT]: [minetweaker.runtime.MTTweaker:load:152]: MineTweaker: Loading PhoenixConfig Botania TC Aspects.zs
    [19:22:02] [Server thread/INFO] [STDOUT]: [minetweaker.runtime.MTTweaker:load:152]: MineTweaker: Loading intermod

I'm thinking the issue lies within the ore dictionary manipulation.

**Comment by UndeadZeratul** [comment]
So, due to some other reasons, we've decided to remove GregTech & IC2 from our pack.  This lowered startup times significantly, as well as possibly reduce server login times enough, even with these lines enabled.  It's a nice coincidence that'll work as a workaround for now.

However,  I also noticed in a few places in our scripts I did something like the following:
```
Ore Dictionary A;
Ore Dictionary B;

A.addAll(B);
A.mirror(B);
```
I noticed that on the wiki page example and in another script of ours, we instead did the following:
```
A.addAll(B);
B.mirror(A);
```
Would this have anything to do with the performance issue?  Or is it simply just the volume of entries that causes eventual slow-down of loading the scripts?
